docs: add slurmctld high availability explanation

dsloanm opened the pull request:
> ~Draft PR as Explanation is not finalized and How-to for deploying HA is still to be added.~ Done

# Pre-submission checklist

 * [X] I read and followed the [CONTRIBUTING guidelines](../CONTRIBUTING.md).
 * [X] I have insured that the documentation tests complete successfully.

## Summary of Changes

[//]: # (Please summarize your commits here. For any complex or contenious changes, please also provide justifications.)

This PR adds documentation to accompany the `slurmctld` high availability functionality currently proposed for the Slurm charms. It includes an Explanation section going into detail on what high availability is and how it is enabled in Charmed HPC, as well as other relevant docs.

#### Related Issues, PRs, and Discussions

[//]: # (Please link to related issues, pull requests, and discussions here - especially corresponding code PRs. If your PR has no related issues, PRs, or discussions, please provide a justification for this PR here instead.)

[feat(slurmctld): implement high availability (HA) for slurmctld #120](https://github.com/charmed-hpc/slurm-charms/pull/120)

dsloanm commented:
Marking this as ready for review. The associated code PR is still on-going but any remaining changes should be fairly minimal.

dsloanm commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,84 @@ Machine  State    Address       Inst id        Base          AZ  Message
+  model_name  = juju_model.slurm.name
+}
+
+resource "juju_integration" "provider-to-filesystem" {
+  model = juju_model.slurm.name
+
+  application {
+    name     = module.filesystem-provider.app_name
+    endpoint = module.filesystem-provider.provides.filesystem
+  }
+
```
Unfortunately, \<angle brackets\> in a Terraform block causes issues with the syntax highlighting, giving a `WARNING: Lexing literal_block ... as "terraform" resulted in an error` message that fails the CI tests. We can't use {braces} or \`backticks\` either.

We could use [square brackets] but that'd overlap with the Terraform syntax for a list. I'm considering all-caps: `FILESYSTEM-PROVIDER` but I think it's too inconsistent with the rest of the docs. What do you think?

dsloanm replied:
I've gone with [square brackets].

dsloanm commented on explanation/high-availability.md:
```diff
@@ -0,0 +1,43 @@
+
+To allow for flexibility in the `StateSaveLocation` shared file system, Charmed HPC implements support for the [`filesystem-client` charm](https://github.com/charmed-hpc/filesystem-charms). This enables users to integrate with the file system of their choice, such as their own CephFS deployment, a cloud-specific managed file system, or another that meets latency requirements. The `slurmctld` char ...
+
+Using the local disk of the `slurmctld` host for `StateSaveLocation` data, without the `filesystem-client`, is supported but limits the HPC cluster to a single `slurmctld` instance. To scale from a single instance to an HA setup, `filesystem-client:mount` must be integrated with `slurmctld:mount`. This performs a one-time migration of `StateSaveLocation` data from the local disk to the shared fil ...
```
Reworded this to make it closer to an Explanation.

AshleyCliff commented on explanation/high-availability.md:
```diff
@@ -0,0 +1,49 @@
+(explanation-slurmctld-high-availability-state-save-location)=
+### Shared `StateSaveLocation` using `filesystem-client` charm
+
+For `slurmctld` HA to function, all `slurmctld` controllers require mounting of the same shared file system to provide a common [`StateSaveLocation`](https://slurm.schedmd.com/slurm.conf.html#OPT_StateSaveLocation) directory for holding controller state data. This directory governs the responsiveness and throughput of the cluster, so should be hosted on a file system with low latency. It is there ...
```
```suggestion
For `slurmctld` HA to function, all `slurmctld` controllers require mounting of the same shared file system to provide a common [`StateSaveLocation`](https://slurm.schedmd.com/slurm.conf.html#OPT_StateSaveLocation) directory to hold controller state data. This directory governs the responsiveness and throughput of the cluster, so it should be hosted on a file system with low latency. It is therefore recommended that the file system not be the same as the file system used on the cluster compute nodes to avoid I/O-intensive user jobs from impacting `slurmctld` responsiveness.
```

AshleyCliff commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,162 @@ Machine  State    Address       Inst id        Base          AZ  Message
+
+The `slurmcltd` charm supports [high availability (HA)](explanation-high-availability) through the native functionality provided by Slurm: an active-passive setup where additional units are backups to a single primary.
+
+This functionality requires a low-latency [shared file system to be deployed](howto-setup-deploy-shared-filesystem) and a `filesystem-client` charm to be integrated with `slurmctld` on the `mount` endpoint to allow sharing of data across all `slurmctld` units. It is recommended that this file system be separate from the file system used for cluster compute nodes to avoid I/O-intensive user jobs f ...
```
There's a statement in the explanation noting that the `filesystem-client` must be deployed without a user-configured mount point - would this be an appropriate place within the how-to to reiterate this point?

AshleyCliff commented on explanation/high-availability.md:
```diff
@@ -0,0 +1,49 @@
+
+In a non-HA setup (a single `slurmctld`), `StateSaveLocation` data is stored on the unit local disk at `/var/lib/slurm/checkpoint`. Before `slurmctld` backup units can be added to enable high availability, the `slurmctld` charm must be integrated with a `filesystem-client` on the `mount` endpoint to provide the necessary shared storage. On integration, the `StateSaveLocation` data is automaticall ...
+
+Once complete, `juju add-unit` can be used to add backup units. It is **not possible to remove the `filesystem-client` integration** and return to a non-HA setup once the migration has completed. To avoid data loss, the files and directories in the local  `/var/lib/slurm/checkpoint` are left untouched following migration.
```
```suggestion
Once complete, `juju add-unit` can be used to add backup units. It is **not possible to remove the `filesystem-client` integration** and return to a non-HA setup once the migration has completed. To avoid data loss, the files and directories in the local  `/var/lib/slurm/checkpoint` are left untouched following migration. Specific steps can be found in the [Migrating a single `slurmctld` to high availability](migrate-slurmctld-high-availability) how-to section.
```

AshleyCliff commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,162 @@ Machine  State    Address       Inst id        Base          AZ  Message
+This is a one-time cluster downtime. Once the data migration is complete, no further downtime is necessary when adding or removing `slurmctld` units.
+:::
+
+To migrate a previously deployed single `slurmctld` unit to an HA setup, integrate a `filesystem-client`, substituting `[filesystem-provider]` with your choice of shared file system provider (file system provider choice is detailed in the [`slurmctld` high availability deployment instructions](deploy-slurmctld-high-availability)). Then scale up the application:
```
```suggestion
To migrate a previously deployed single `slurmctld` unit to an HA setup, integrate a `filesystem-client` following the file system deployment process described in the [`slurmctld` high availability section](deploy-slurmctld-high-availability), substituting `[filesystem-provider]` with your choice of shared file system provider. Then scale up the application:
```

AshleyCliff commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,162 @@ Machine  State    Address       Inst id        Base          AZ  Message
+::::
+
+:::::
+
```
A closing statement and example on what the scaled up system would look like with the appropriate commands would be helpful here

dsloanm replied:
Added this. Since this is a specific example, I included CephFS as the shared file system since that's what we've been using in testing.

I also went back and changed the other `juju status` outputs to include the new `primary - UP` unit message for `slurmcltd`. It's now present even in a non-HA setup.

AshleyCliff reviewed (changes requested):
Looking good - the migrating section is a useful addition! A couple wording tweaks and a few clarifying suggestions and then we'll be good to go!

dsloanm commented:
Latest commits add some bold text to stress use of a separate HA file system and more emphatically spell out an HA file system should be provisioned and then a separate compute node file system.

I also took inspiration from [the OpenStack docs](https://canonical-openstack.readthedocs-hosted.com/en/latest/how-to/install/install-canonical-openstack-using-the-manual-bare-metal-provider/#related-how-to-guides) and added a "Related how-to guides" section to the end of the "Deploy Slurm" page. This was instead of "Next steps" - that felt more tutorial-y to me.

I've changed the file system prerequisites to just "A Slurm cluster" from "A deployed Slurm cluster" to slightly loosen the implied ordering. I've avoided explicitly mentioning slurmctld HA in the file system how-to just for separation of concerns.

AshleyCliff commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,82 @@ Machine  State    Address       Inst id        Base          AZ  Message
+juju deploy filesystem-client --channel latest/edge
+juju integrate filesystem-client:filesystem [filesystem-provider]:filesystem
+
+juju deploy slurmctld --base "ubuntu@24.04" --channel "edge" --num-units 2
```
What does this `juju deploy slurmctld` do to the already deployed slurmctld controller from above in the main 'Deploy Slurm' section?

dsloanm replied:
Good catch - if there's a `slurmctld` already deployed, doing another `juju deploy slurmctld` would either fail with an "application already exists" error or deploy a brand new application, if either `deploy` used a different name.

I've added a note to the instructions above to read the high availability section instead if HA is needed.

AshleyCliff commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,82 @@ Machine  State    Address       Inst id        Base          AZ  Message
+
+The `slurmcltd` charm optionally supports [high availability (HA)](explanation-high-availability) through the native functionality provided by Slurm: an active-passive setup where additional units are backups to a single primary.
+
+This functionality requires a low-latency [shared file system to be deployed](howto-setup-deploy-shared-filesystem) and a `filesystem-client` charm, without a user-configured mount point, to be integrated with `slurmctld` on the `mount` endpoint to allow sharing of data across all `slurmctld` units. For guidance on choosing a file system, see the [Shared `StateSaveLocation` using `filesystem-clie ...
+
+It is recommended that the HA file system **not be the same as the file system used for the cluster compute nodes** to avoid I/O-intensive user jobs from impacting `slurmctld` responsiveness. The suggested approach is to deploy a dedicated HA file system then subsequently provision a separate file system for the compute nodes.
+
+Once a chosen shared file system has been deployed and made available via a proxy or other file system provider charm, run the following, substituting `[filesystem-provider]` with the name of the provider charm, to deploy a `slurmctld` HA setup with two units (a primary and single backup):
```
```suggestion
This functionality requires a low-latency shared file system to be deployed and a `filesystem-client` charm to be integrated with `slurmctld`. For guidance on choosing a file system, see the [Shared `StateSaveLocation` using `filesystem-client` charm](explanation-slurmctld-high-availability-state-save-location) section. It is recommended that the HA file system **not be the same as the file system used for the cluster compute nodes** to avoid I/O-intensive user jobs from impacting `slurmctld` responsiveness. The suggested approach is to deploy a dedicated HA file system then subsequently provision a separate file system for the compute nodes.

To deploy the file system, follow the instructions in the [Deploy a shared filesystem](howto-setup-deploy-shared-filesystem) section. Once a chosen shared file system has been deployed and made available via a proxy or provider charm, run the following, substituting `[filesystem-provider]` with the name of the provider charm, to deploy a `slurmctld` HA setup with two units (a primary and single backup):
```

dsloanm replied:
Fixed!

AshleyCliff commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,82 @@ Machine  State    Address       Inst id        Base          AZ  Message
+::::
+
+:::::
+
```
Might be worth either showing `juju status` here like it's done in the manage section now, or potentially just a quick note to check `juju status` and note the number of `slurmctld` units.

dsloanm replied:
Agreed, I've copied the `juju status` output from the manage section.

dsloanm commented on explanation/high-availability.md:
```diff
@@ -0,0 +1,55 @@
+
+Similarly to `StateSaveLocation`, data in `/etc/slurm` is migrated to `/srv/slurmctld-statefs/etc/slurm` on `filesystem-client` integration. The `/etc/slurm` directory is then replaced with a symbolic link to `/srv/slurmctld-statefs/etc/slurm` on all `slurmctld` instances to ensure all access the same configuration files.
+
+To avoid data loss, any existing `/etc/slurm/` is backed up to a date-stamped directory on the unit's local disk, for example `/etc/slurm_20250620_161437` for a backup performed on 2025-06-20 at 16:14:37. To prevent non-leaders from reading partially written configuration files, updates to files are made atomically via [slurmutils](https://github.com/charmed-hpc/slurmutils/).
```
"Atomically" in the sense of an ["atomic operation"](https://wiki.osdev.org/Atomic_operation) or an ["atomic transaction"](https://en.wikipedia.org/wiki/Atomicity_(database_systems)), i.e. it's either "done" or "not done", it can't be "partially done". Here, it means a non-leader will always get a complete `slurm.conf` - it won't crash or malfunction trying to read a partially done `slurm.conf`. If the leader is in the middle of writing a new `slurm.conf`, any read from the non-leader will see the complete, old `slurm.conf`. The non-leader can only read the new `slurm.conf` after the leader has completely finished writing.

To accomplish this, in `slurmutils` we write to `.slurm.conf.swp` then [`rename`](https://www.man7.org/linux/man-pages/man2/rename.2.html#DESCRIPTION) it to `slurm.conf`, meaning anyone reading `slurm.conf` will always see a complete file. This is guaranteed by the `rename` operation: "If newpath already exists, it will be atomically replaced, so that there is no point at which another process attempting to access newpath will find it missing."

dsloanm commented on howto/setup/deploy-slurm.md:
```diff
@@ -439,6 +439,82 @@ Machine  State    Address       Inst id        Base          AZ  Message
+(deploy-slurmctld-high-availability)=
+### Deploying `slurmctld` in high availability
+
+The `slurmcltd` charm optionally supports [high availability (HA)](explanation-high-availability) through the native functionality provided by Slurm: an active-passive setup where additional units are backups to a single primary.
```
I like this one as "controllers" since it refers to native Slurm functionality and not Juju.

dsloanm commented:
The latest commits address the most recent comments. I think the only sticking point is "controller" versus "unit". I've gone with "unit" where speaking specifically about Juju-managed `slurmctld` instances and "controller" when referring to the native Slurm component, in which case I say we're OK keeping the Manage Slurm section title as-is.

dsloanm commented:
That last suggestion sounds good.

I've slightly reworded it to "Charmed HPC uses Slurm's built-in HA support alongside [Juju's horizontal scaling capabilities](https://documentation.ubuntu.com/juju/latest/reference/scaling/) to allow for the `slurmctld` charm to run with multiple units in an HA configuration" since I think "flexible" sounded slightly ambiguous.

I also tweaked the paragraph above from "The `slurmctld` controller service supports HA through an active-passive setup" to "Slurm's `slurmctld` controller service has built-in HA support through an active-passive setup" to make it clear where the HA support is coming from.

AshleyCliff reviewed (approved):
Looks good, we're ready to merge!

NucciTheBoss commented on howto/setup/deploy-slurm.md:
```diff
@@ -555,3 +681,9 @@ To verify that the entire partition is `IDLE`, run `sinfo`{l=shell} without the
 juju exec -u sackd/0 -- sinfo
 :::
+
+## Related how-to guides
+
+Now that Slurm is deployed, you might want to view:
+
+- {ref}`howto-setup-deploy-shared-filesystem`
```
We're currently using "Next Steps" to indicate where the user should go next such as in the [COS how-to](https://canonical-charmed-hpc.readthedocs-hosted.com/latest/howto/manage/integrate-with-cos/), so for consistency, I think we should use "Next Steps" here instead

```suggestion
## Next Steps

Now that Slurm is deployed, you might want to view:

- {ref}`howto-setup-deploy-shared-filesystem`
```

The other benefit of using "Next Steps" is that it doesn't limit the scope of where we can tell the user to go next. For example, we could eventually have two paragraphs here, where one is like "here's the next steps for deployment", and the other is "here's some reference documentation to help better understand the architecture of Slurm"

NucciTheBoss replied:
Also, not sure how I feel "you might" here for the shared filesystem. It's a must for an actual working HPC cluster :sweat_smile: 

We could say something along the lines here like "now that Slurm is deployed, it is time to deploy the shared filesystem of your Charmed HPC cluster". "You might" here implies to me that this step is optional when it really isn't.

NucciTheBoss reviewed (changes requested):
Just two comments about the "Next Steps" section in the "Deploy Slurm" how-to. Once those are addressed, this PR is good to go!

NucciTheBoss reviewed (approved):
LGTM! :shipit:

NucciTheBoss commented:
The documentation checks are taking 10+ minutes because it's currently being rate limited by GitHub, so I am going to go ahead and merge this PR into `main`. Nice work @dsloanm!
